### PR TITLE
Makes it possible to attack resin doors on help intent

### PR DESCRIPTION
**Pull request by Xander3359**
## About The Pull Request
You can now attack resin walls on help intent
## Why It's Good For The Game
Gameplay consistency
## Changelog
:cl:
fix: You can now attack resin walls on help intent
/:cl:

### CODE REVIEW

**Comment by Naaanii** [comment]
Based but please bring back auto click. My mouse life expectancy has gone up centuries when it was TMd.

File: code/game/objects/structures/mineral_doors.dm [hunk 1]
```diff
@@ -83,26 +83,22 @@
...
+		return TRUE
 	if(QDELETED(src))
 		return
+	if(istype(attacking_item, /obj/item/tool/pickaxe/plasmacutter) && !user.do_actions) //Special plasmacutter interaction
+		var/obj/item/tool/pickaxe/plasmacutter/pcutter = attacking_item
+		if(pcutter.start_cut(user, name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD))
+			pcutter.cut_apart(user, name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD) //Minimal energy cost.
+			user.changeNext_move(attacking_item.attack_speed)
+			user.do_attack_animation(src, used_item = attacking_item)
+			take_damage(src.max_integrity, BURN, FIRE, armour_penetration = 100)
```
Comment (QualityVan) [inline code/game/objects/structures/mineral_doors.dm:98]: src., also balance change

**Comment by QualityVan** [comment]
Freeze.

**Comment by Xander3359** [comment]
I didn't consider this a balance change really but... alright.

**Comment by coldironwarrior** [comment]
joever for arthritics everywhere

**Comment by QualityVan** [comment]
(slightly delayed) Unfreeze.

File: code/game/objects/structures/mineral_doors.dm [hunk 2]
```diff
@@ -83,26 +83,22 @@
 		icon_state = "[base_icon_state][smoothing_flags ? "-[smoothing_junction]" : ""]"
 
 
-/obj/structure/mineral_door/attackby(obj/item/W, mob/living/user)
+/obj/structure/mineral_door/attackby(obj/item/attacking_item, mob/living/user)
```
Comment (QualityVan) [inline code/game/objects/structures/mineral_doors.dm:86]: Given that this is explicitly a pr to change resin doors, why are the changes on mineral doors as a whole?

File: code/game/objects/structures/mineral_doors.dm (hunk 1 above)
Comment (QualityVan) [inline code/game/objects/structures/mineral_doors.dm]: Why the change from acting as a hit with a bonus multiplier to instant deletion for all walls, and why the armor pen?

Comment (Xander3359) [reply code/game/objects/structures/mineral_doors.dm]: Done

File: code/game/objects/structures/mineral_doors.dm [hunk 3]
```diff
@@ -83,26 +83,22 @@
...
-			if(istype(src, /obj/structure/mineral_door/resin))
-				multiplier += PLASMACUTTER_RESIN_MULTIPLIER //Plasma cutters are particularly good at destroying resin structures.
-			else
-				multiplier += PLASMACUTTER_RESIN_MULTIPLIER * 0.5
-			P.cut_apart(user, src.name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD) //Minimal energy cost.
-	if(W.damtype == BURN && istype(src, /obj/structure/mineral_door/resin)) //Burn damage deals extra vs resin structures (mostly welders).
-		multiplier += 1 //generally means we do double damage to resin doors
-
-	take_damage(max(0, W.force * multiplier - W.force), W.damtype, MELEE)
+	attacking_item.attack_obj(src, user)
```
Comment (QualityVan) [inline code/game/objects/structures/mineral_doors.dm:103]: This makes attack_obj run twice if it returns false. Look at how obj/alien does it.

File: code/game/objects/structures/mineral_doors.dm [hunk 4]
```diff
@@ -83,26 +83,22 @@
...
+		return TRUE
 	if(QDELETED(src))
 		return
+	if(istype(attacking_item, /obj/item/tool/pickaxe/plasmacutter) && !user.do_actions) //Special plasmacutter interaction
+		var/obj/item/tool/pickaxe/plasmacutter/pcutter = attacking_item
+		if(pcutter.start_cut(user, name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD))
+			pcutter.cut_apart(user, name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD) //Minimal energy cost.
+			user.changeNext_move(attacking_item.attack_speed)
+			user.do_attack_animation(src, used_item = attacking_item)
+			take_damage(max_integrity, BURN, FIRE, armour_penetration = 100)
```
Comment (QualityVan) [inline code/game/objects/structures/mineral_doors.dm:100]: Why the change from a normal hit from the item with multipliers to just instant deletion, ignoring all armor? This also removes other burn-based melee weapons (small niche, admittedly) from having a bonus.

**Review by QualityVan — CHANGES_REQUESTED** [review]
There's one extra comment since the line it was on disappeared from a commit made while I was in progress, oops.

File: code/game/objects/structures/mineral_doors.dm (hunk 4 above)
Comment (Xander3359) [reply code/game/objects/structures/mineral_doors.dm:100]: Because plasma cutter instantly cuts wall, so it should cut doors as well.

Comment (ivanmixo) [reply code/game/objects/structures/mineral_doors.dm:99]: That is besides the point since this also affects things that arent a plasmacutter

**Comment by Lumipharon** [comment]
What they said. What this PR is labelled as is not what its actually doing.

File: code/_onclick/item_attack.dm [hunk 5]
```diff
@@ -67,9 +67,6 @@
 	if(.)
 		return TRUE
 
-	if(user.a_intent != INTENT_HARM)
-		return
```
Comment (Lumipharon) [inline code/_onclick/item_attack.dm:71]: You have not thought this through at all. Why are you changing this at the object level.

**Review by Lumipharon — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/mineral_doors.dm [hunk 6]
```diff
@@ -83,26 +83,30 @@
...
 	. = ..()
 	if(.)
 		return
 	if(QDELETED(src))
 		return
 
+	calculate_attackby(attacking_item, user)
+
+///Handles damage calculation + plasmacutter interaction from attackby
+/obj/structure/mineral_door/proc/calculate_attackby(obj/item/attacking_item, mob/living/user)
```
Comment (Lumipharon) [inline code/game/objects/structures/mineral_doors.dm:96]: this isn't very clear. This isn't a calculation proc, its doing actual damage, specifically if an item is doing burn damage.

File: code/game/objects/structures/mineral_doors.dm [hunk 7]
```diff
@@ -83,26 +83,30 @@
...
 				multiplier += PLASMACUTTER_RESIN_MULTIPLIER //Plasma cutters are particularly good at destroying resin structures.
 			else
 				multiplier += PLASMACUTTER_RESIN_MULTIPLIER * 0.5
 			P.cut_apart(user, src.name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD) //Minimal energy cost.
-	if(W.damtype == BURN && istype(src, /obj/structure/mineral_door/resin)) //Burn damage deals extra vs resin structures (mostly welders).
+	if(attacking_item.damtype == BURN && istype(src, /obj/structure/mineral_door/resin)) //Burn damage deals extra vs resin structures (mostly welders).
 		multiplier += 1 //generally means we do double damage to resin doors
 
-	take_damage(max(0, W.force * multiplier - W.force), W.damtype, MELEE)
+	take_damage(max(0, attacking_item.force * multiplier - attacking_item.force), attacking_item.damtype, MELEE)
```
Comment (Lumipharon) [inline code/game/objects/structures/mineral_doors.dm:109]: You can early return the entire proc if its not a PC and doesn't have burn damage.

File: code/game/objects/structures/mineral_doors.dm [hunk 8]
```diff
@@ -83,26 +83,30 @@
...
 
-/obj/structure/mineral_door/attackby(obj/item/W, mob/living/user)
+/obj/structure/mineral_door/attackby(obj/item/attacking_item, mob/living/user)
 	. = ..()
 	if(.)
 		return
 	if(QDELETED(src))
 		return
 
+	calculate_attackby(attacking_item, user)
```
Comment (Lumipharon) [inline code/game/objects/structures/mineral_doors.dm:93]: the way this is structured, this won't be called if you ARE in harm intent for non resin type mineral walls.

File: code/_onclick/item_attack.dm [hunk 9]
```diff
@@ -71,7 +71,7 @@
 		return
 
 	if(obj_flags & CAN_BE_HIT)
-		return I.attack_obj(src, user)
+		return attacking_item.attack_obj(src, user)
```
Comment (Lumipharon) [inline code/_onclick/item_attack.dm:74]: I think you're complicating the proc chain here, although thats probabably because it was bad in the first place.
attackby calls attack_obj here, which in turn calls the extremely well named attacked_by, where the actual damage is done.
Rather than this confusing shit, all the bonus damage stuff should be handed in attacked_by.

Then all you really need to do is call attacking_item.attack_obj(src, user) again if we don't already have a return value (and checking CAN_BE_HIT again I guess)

**Review by Lumipharon — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/mineral_doors.dm [hunk 10]
```diff
@@ -83,26 +83,47 @@
 		icon_state = "[base_icon_state][smoothing_flags ? "-[smoothing_junction]" : ""]"
 
 
-/obj/structure/mineral_door/attackby(obj/item/W, mob/living/user)
+/obj/structure/mineral_door/attackby(obj/item/attacking_item, mob/living/user)
 	. = ..()
 	if(.)
-		return
+		return TRUE
```
Comment (Lumipharon) [inline code/game/objects/structures/mineral_doors.dm:89]: don't do this. if there is already a return value, we return that. you're overriding it. While in this case that probably doesn't matter, its a bad habit.

File: code/game/objects/structures/mineral_doors.dm [hunk 11]
```diff
@@ -83,26 +83,47 @@
...
-				multiplier += PLASMACUTTER_RESIN_MULTIPLIER //Plasma cutters are particularly good at destroying resin structures.
-			else
-				multiplier += PLASMACUTTER_RESIN_MULTIPLIER * 0.5
-			P.cut_apart(user, src.name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD) //Minimal energy cost.
-	if(W.damtype == BURN && istype(src, /obj/structure/mineral_door/resin)) //Burn damage deals extra vs resin structures (mostly welders).
-		multiplier += 1 //generally means we do double damage to resin doors
+	if(user.a_intent == INTENT_HARM)
+		return
+
+	if(obj_flags & CAN_BE_HIT)
```
Comment (Lumipharon) [inline code/game/objects/structures/mineral_doors.dm:96]: early return

File: code/game/objects/structures/mineral_doors.dm [hunk 12]
```diff
@@ -83,26 +83,47 @@
...
+	. = ..()
+	if(attacking_item.damtype != BURN)
+		return
+	var/damage_multiplier = take_extra_burn_damage(attacking_item, user, def_zone)
+
+	take_damage(max(0, attacking_item.force * damage_multiplier - attacking_item.force), attacking_item.damtype, MELEE)
+
+///Takes extra damage if our attacking item does burn damage
+/obj/structure/mineral_door/proc/take_extra_burn_damage(obj/item/attacking_item, mob/living/user, def_zone, bonus_damage = 1)
+	if(!isplasmacutter(attacking_item) || !user.do_actions)
```
Comment (Lumipharon) [inline code/game/objects/structures/mineral_doors.dm:109]: what is this do_actions check

File: code/game/objects/structures/mineral_doors.dm [hunk 13]
```diff
@@ -83,26 +83,47 @@
...
+
+	var/obj/item/tool/pickaxe/plasmacutter/attacking_pc = attacking_item
+	if(attacking_pc.start_cut(user, name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD, no_string = TRUE))
+		bonus_damage += PLASMACUTTER_RESIN_MULTIPLIER * 0.5
+		attacking_pc.cut_apart(user, src.name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD) //Minimal energy cost.
+
+	return bonus_damage
+
+/obj/structure/mineral_door/resin/take_extra_burn_damage(obj/item/attacking_item, mob/living/user, def_zone, bonus_damage = 1)
+	bonus_damage = ..()
```
Comment (Lumipharon) [inline code/game/objects/structures/mineral_doors.dm:120]: this parent return is kinda gross, although this whole thing is pretty wew.
you're checking for PC again but not the cutapart check, which means it could fail in parent but you're still applying a half bonus here.

**Review by Lumipharon — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/mineral_doors.dm (hunk 12 above)
Comment (Lumipharon) [reply code/game/objects/structures/mineral_doors.dm:111]: unresolved. also this proc is incorrectly named

File: code/game/objects/structures/mineral_doors.dm [hunk 14]
```diff
@@ -83,26 +83,48 @@
...
+	if(attacking_pc.start_cut(user, name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD, no_string = TRUE))
+		bonus_damage += PLASMACUTTER_RESIN_MULTIPLIER * 0.5
+		attacking_pc.cut_apart(user, src.name, src, PLASMACUTTER_BASE_COST * PLASMACUTTER_VLOW_MOD) //Minimal energy cost.
+
+	return bonus_damage
+
+/obj/structure/mineral_door/resin/take_extra_burn_damage(obj/item/attacking_item, mob/living/user, def_zone, bonus_damage = 1)
+	. = ..()
+	bonus_damage = .
+	if(bonus_damage <= 1)
```
Comment (Lumipharon) [inline code/game/objects/structures/mineral_doors.dm:124]: no. this is smelly and lazy, and relies on parent not changing in anyway in the future.

**Review by Lumipharon — CHANGES_REQUESTED** [review]
this still has issues

**Review by Lumipharon — APPROVED** [review]
ok this is still aids and pc stuff should be a pc proc, but this is fine for now.
